File: python/valuecell/server/api/schemas/homepage_context.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class HomepageSignalData(BaseModel):
    label: str = Field(..., description="Signal label")
    value: Any = Field(..., description="Signal value")


class HomepageStagePointData(BaseModel):
    trading_date: str = Field(..., description="Trading date")
    cycle_stage: str = Field(..., description="Cycle stage")
    stage_score: int = Field(..., description="Stage score")


class HomepageTurningPointData(BaseModel):
    trading_date: str = Field(..., description="Trading date")
    from_stage: str = Field(..., description="Previous stage")
    to_stage: str = Field(..., description="Current stage")


class HomepageMarketOverviewData(BaseModel):
    available: bool = Field(..., description="Whether market snapshot is available")
    market_state: Optional[str] = Field(None, description="Market state")
    summary: Optional[str] = Field(None, description="Readable market summary")
    confidence: Optional[str] = Field(None, description="Confidence level")
    action_hint: Optional[str] = Field(None, description="Action hint")
    signals: list[HomepageSignalData] = Field(default_factory=list)
    score: Optional[int] = Field(None, description="Market score")
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageEmotionCycleData(BaseModel):
    available: bool = Field(..., description="Whether emotion cycle is available")
    cycle_stage: Optional[str] = Field(None, description="Current cycle stage")
    summary: Optional[str] = Field(None, description="Readable cycle summary")
    action_hint: Optional[str] = Field(None, description="Action hint")
    trend_direction: Optional[str] = Field(None, description="Recent direction")
    stage_points: list[HomepageStagePointData] = Field(default_factory=list)
    turning_points: list[HomepageTurningPointData] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageThemeInstitutionData(BaseModel):
    source: str = Field(..., description="Institution proxy source")
    net_inflow: float = Field(..., description="Net inflow amount")


class HomepageThemeMetricsData(BaseModel):
    score: int = Field(..., description="Theme score")
    change_value: float = Field(..., description="Theme daily change")
    ths_flow_value: float = Field(..., description="THS flow")
    dc_flow_value: float = Field(..., description="DC flow")
    kpl_count: int = Field(..., description="Short-term co-move count")
    hot_count: int = Field(..., description="Hot list count")


class HomepageThemeItemData(BaseModel):
    trading_date: Optional[str] = Field(None, description="Trading date")
    theme_name: str = Field(..., description="Theme name")
    theme_code: str = Field(..., description="Theme code")
    theme_state: str = Field(..., description="Theme state")
    summary: str = Field(..., description="Theme summary")
    representative_tickers_json: list[str] = Field(default_factory=list)
    rank: int = Field(..., description="Theme rank")
    expectation_gap_level: str = Field(..., description="Expectation gap level")
    core_leaders_json: list[str] = Field(default_factory=list)
    core_institutions_json: list[HomepageThemeInstitutionData] = Field(
        default_factory=list
    )
    metrics: HomepageThemeMetricsData = Field(..., description="Theme metrics")


class HomepageThemeFocusData(BaseModel):
    available: bool = Field(..., description="Whether theme focus is available")
    items: list[HomepageThemeItemData] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageActionFrameworkData(BaseModel):
    available: bool = Field(..., description="Whether action framework is available")
    title: str = Field(..., description="Section title")
    summary: Optional[str] = Field(None, description="Framework summary")
    focus_points: list[str] = Field(default_factory=list)
    avoid_points: list[str] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageWatchlistObservationItemData(BaseModel):
    ticker: str = Field(..., description="Ticker")
    display_name: str = Field(..., description="Display name")
    watchlist_name: str = Field(..., description="Watchlist name")
    price: Optional[str] = Field(None, description="Formatted price")
    change_percent: Optional[float] = Field(None, description="Change percent")
    status: str = Field(..., description="Observation status")
    reason: str = Field(..., description="Observation reason")
    theme_name: Optional[str] = Field(None, description="Related theme name")


class HomepageWatchlistObservationData(BaseModel):
    available: bool = Field(..., description="Whether watchlist observation is available")
    items: list[HomepageWatchlistObservationItemData] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageMetricData(BaseModel):
    label: str = Field(..., description="Metric label")
    value: str = Field(..., description="Metric value")


class HomepagePortfolioHandlingData(BaseModel):
    available: bool = Field(..., description="Whether portfolio handling is available")
    summary: Optional[str] = Field(None, description="Summary")
    holding_count: int = Field(..., description="Holding count")
    focus_count: int = Field(..., description="Focus holding count")
    action_breakdown: list[HomepageMetricData] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageRiskControlData(BaseModel):
    available: bool = Field(..., description="Whether risk control section is available")
    summary: str = Field(..., description="Risk control summary")
    position_suggestion: str = Field(..., description="Position suggestion")
    signals: list[str] = Field(default_factory=list)
    empty_message: Optional[str] = Field(None, description="Fallback message")


class HomepageContextData(BaseModel):
    generated_at: datetime = Field(..., description="Generation timestamp")
    market_overview: HomepageMarketOverviewData = Field(...)
    emotion_cycle: HomepageEmotionCycleData = Field(...)
    theme_focus: HomepageThemeFocusData = Field(...)
    action_framework: HomepageActionFrameworkData = Field(...)
    watchlist_observation: HomepageWatchlistObservationData = Field(...)
    portfolio_handling: HomepagePortfolioHandlingData = Field(...)
    risk_control: HomepageRiskControlData = Field(...)

File: frontend/src/types/homepage-context.ts
export type HomepageSignal = {
  label: string;
  value: string | number | null;
};

export type HomepageStagePoint = {
  trading_date: string;
  cycle_stage: string;
  stage_score: number;
};

export type HomepageTurningPoint = {
  trading_date: string;
  from_stage: string;
  to_stage: string;
};

export type HomepageMarketOverview = {
  available: boolean;
  market_state: string | null;
  summary: string | null;
  confidence: string | null;
  action_hint: string | null;
  signals: HomepageSignal[];
  score: number | null;
  empty_message: string | null;
};

export type HomepageEmotionCycle = {
  available: boolean;
  cycle_stage: string | null;
  summary: string | null;
  action_hint: string | null;
  trend_direction: string | null;
  stage_points: HomepageStagePoint[];
  turning_points: HomepageTurningPoint[];
  empty_message: string | null;
};

export type HomepageThemeInstitution = {
  source: string;
  net_inflow: number;
};

export type HomepageThemeMetrics = {
  score: number;
  change_value: number;
  ths_flow_value: number;
  dc_flow_value: number;
  kpl_count: number;
  hot_count: number;
};

export type HomepageThemeItem = {
  trading_date: string | null;
  theme_name: string;
  theme_code: string;
  theme_state: string;
  summary: string;
  representative_tickers_json: string[];
  rank: number;
  expectation_gap_level: string;
  core_leaders_json: string[];
  core_institutions_json: HomepageThemeInstitution[];
  metrics: HomepageThemeMetrics;
};

export type HomepageThemeFocus = {
  available: boolean;
  items: HomepageThemeItem[];
  empty_message: string | null;
};

export type HomepageActionFramework = {
  available: boolean;
  title: string;
  summary: string | null;
  focus_points: string[];
  avoid_points: string[];
  empty_message: string | null;
};

export type HomepageWatchlistObservationItem = {
  ticker: string;
  display_name: string;
  watchlist_name: string;
  price: string | null;
  change_percent: number | null;
  status: string;
  reason: string;
  theme_name: string | null;
};

export type HomepageWatchlistObservation = {
  available: boolean;
  items: HomepageWatchlistObservationItem[];
  empty_message: string | null;
};

export type HomepageMetric = {
  label: string;
  value: string;
};

export type HomepagePortfolioHandling = {
  available: boolean;
  summary: string | null;
  holding_count: number;
  focus_count: number;
  action_breakdown: HomepageMetric[];
  empty_message: string | null;
};

export type HomepageRiskControl = {
  available: boolean;
  summary: string;
  position_suggestion: string;
  signals: string[];
  empty_message: string | null;
};

export type HomepageContext = {
  generated_at: string;
  market_overview: HomepageMarketOverview;
  emotion_cycle: HomepageEmotionCycle;
  theme_focus: HomepageThemeFocus;
  action_framework: HomepageActionFramework;
  watchlist_observation: HomepageWatchlistObservation;
  portfolio_handling: HomepagePortfolioHandling;
  risk_control: HomepageRiskControl;
};
